Runtime/Scripts/DebugUtils/RGGizmos.cs

```diff
@@ -0,0 +1,396 @@
+            {
+                Debug.DrawLine(lineParams.Item1, lineParams.Item2, lineParams.Item3);
+            }
+            
+            // Draw spheres
+            foreach (var sphereParams in _spheresAtPosition.Values)
+            {
+                if (sphereParams.Item4)
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawWireSphere(sphereParams.Item1, sphereParams.Item3);
+                }
+                else
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawSphere(sphereParams.Item1, sphereParams.Item3);
+                }
+            }
+            foreach (var sphereParams in _spheresAtEntity.Values)
+            {
+                var originInstance = RGFindUtils.Instance.FindOneByInstanceId<RGEntity>(sphereParams.Item1);
+                if (originInstance == null) continue;
+                if (sphereParams.Item4)
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawWireSphere(originInstance.transform.position, sphereParams.Item3);
+                }
+                else
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawSphere(originInstance.transform.position, sphereParams.Item3);
+                }
+            }
+            
+            // Delete any billboards that are no longer being drawn
+            var billboardsToRemove = _drawnBillboards.Where(b => !_billboardsToDraw.ContainsKey(b.Key));
+            foreach (var billboard in billboardsToRemove)
+            {
+                Object.Destroy(billboard.Value);
+                _drawnBillboards.Remove(billboard.Key, out _);
+            }
+            
+            // Now update or create any billboards that are being drawn
+            foreach (var billboardParams in _billboardsToDraw)
+            {
+                try
+                {
+                    // If the billboard does not exist, create it
+                    if (!_drawnBillboards.ContainsKey(billboardParams.Key))
+                    {
+                        var entityGameObject = RGFindUtils.Instance.FindOneByInstanceId<RGEntity>(billboardParams.Key);
+                        if (entityGameObject == null) continue;
+                        var billboardObject = Object.Instantiate(_billboardAsset, entityGameObject.transform.position,
+                            Quaternion.identity);
+                        billboardObject.transform.SetParent(entityGameObject.transform);
+                        _drawnBillboards[billboardParams.Key] = billboardObject;
+                    }
+                
+                    // Then set the parameters
+                    var billboard = _drawnBillboards[billboardParams.Key];
+                    if (billboard == null || !billboard.activeInHierarchy) continue;
```

Comment:
Here's another bit of funky C# pattern matching for you to consider. Again, a style thing, so I'm fine if it feels less clear. This could be:

```suggestion
                    if (billboard is not { activeInHierarchy: true }) continue;
```

This pattern breaks down as:

* `billboard is not ...` - Only call the `if` block if `billboard` does **not** match the specified pattern
* `{ activeInHierarchy: true }` - `billboard` must be a non-`null` object **with** the `activeInHierarchy` property set to `true`.

It does read fairly well to me. You're saying "If the billboard is not `activeInHierarchy`, then `continue`". A `null` billboard is clearly not `activeInHierarchy`.

Reply:
Looks better to me as well! Added.

Reply:
Ok, given the above stuff about "bypassing lifetime access checks", I think you should return to the original code path here. The `== null` comparison has special meaning in Unity and I'd prefer we not rock the boat and miss something important.

Reply:
Reverted!